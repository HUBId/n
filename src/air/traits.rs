//! Traits und Vertragsstrukturen der AIR-Schicht.
//!
//! Die hier definierten Traits beschreiben die Schnittstelle zwischen der
//! Spezifikation und den spaeteren Implementierungen. Es werden keine
//! Evaluierungen vorgenommen; alle Rueckgaben sind deskriptiv.

use super::context::AirContext;
use super::descriptors::{
    BoundaryConstraintDescriptor, CompositionPolynomialDescriptor, LookupArgumentDescriptor,
    PermutationArgumentDescriptor,
};
use super::ids::AirSpecId;
use super::inputs::PublicInputs;
use super::selectors::SelectorSet;

/// Gruppentyp eines Registers.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TraceGroup {
    /// Kernregister mit arithmetischen Zuständen.
    Core,
    /// Hilfsregister fuer Carry-Bits, Range-Beweise etc.
    Auxiliary,
}

/// Reihenfolgeeintrag fuer eine Uebergangsbedingung.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TransitionConstraintOrder {
    /// Name der Constraint-Klasse.
    pub name: &'static str,
    /// Gruppenzugehörigkeit des Constraint-Vektors.
    pub group: TraceGroup,
    /// Index innerhalb der kanonischen Liste.
    pub index: usize,
    /// Optionaler Selektor, der die Bedingung aktiviert.
    pub selector_column: Option<usize>,
}

/// Beschreibung, welche Constraints bei der Auswertung einer Zeile relevant sind.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TransitionEvaluationDescriptor {
    /// Zeilenindex innerhalb der dichten Spur.
    pub row_index: usize,
    /// Reihenfolge der aktiven Constraints.
    pub active_constraints: Vec<TransitionConstraintOrder>,
}

/// Beschreibung der Randbedingungen fuer erste und letzte Zeile.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BoundaryEvaluationDescriptor {
    /// Constraints, die auf der ersten Zeile gelten.
    pub first_row: Vec<BoundaryConstraintDescriptor>,
    /// Constraints, die auf der letzten Zeile gelten.
    pub last_row: Vec<BoundaryConstraintDescriptor>,
}

/// Beschreibung einer Lookup-Tabelle innerhalb der AIR.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LookupTableAccess {
    /// Name der Tabelle.
    pub name: &'static str,
    /// Digest der Tabellenparameter.
    pub table_digest: super::ids::ParameterDigest,
    /// Lookup-Argument, das gegen diese Tabelle geprüft wird.
    pub argument: LookupArgumentDescriptor,
}

/// Beschreibung der Zuordnung von Transcript-Challenges zu Constraint-Gruppen.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CompositionChallengeSet {
    /// Indizes der Boundary-Alphas entsprechend `boundary_order`.
    pub boundary_alphas: Vec<usize>,
    /// Indizes der Transition-Alphas entsprechend `transition_order`.
    pub transition_alphas: Vec<usize>,
    /// Zufallswerte fuer Randomizer-Polynome nach Registergruppe.
    pub randomizer_alphas: Vec<usize>,
}

/// Trait, das jede konkrete AIR-Instanz implementieren muss.
pub trait AirSpec {
    /// Zugehoeriger Public-Input-Typ.
    type PublicInputs: PublicInputs;

    /// Eindeutige Kennung der Spezifikation.
    fn id(&self) -> AirSpecId;
    /// Kontextinformationen fuer Gradgrenzen, Spur und Domains.
    fn context(&self) -> &AirContext;
    /// Typisierte Sicht auf die Public Inputs.
    fn public_inputs(&self) -> &Self::PublicInputs;
    /// Beschreibung der Uebergangsevaluierung fuer eine Zeile.
    fn evaluate_transition(&self, row: usize) -> TransitionEvaluationDescriptor;
    /// Beschreibung der Boundary-Constraints fuer erste/letzte Zeile.
    fn evaluate_boundary(&self) -> BoundaryEvaluationDescriptor;
    /// Optionale Lookup-Tabellen (Range/Opcode/etc.).
    fn lookup_tables(&self) -> &[LookupTableAccess];
    /// Selektorwerte in kanonischer Reihenfolge.
    fn selectors(&self) -> &SelectorSet;
    /// Ordnung fuer das Kompositionspolynom.
    fn composition_descriptor(&self) -> &CompositionPolynomialDescriptor;
    /// Zuordnung der Transcript-Challenges.
    fn composition_challenges(&self) -> CompositionChallengeSet;
    /// Beschreibung der verwendeten Permutationsargumente.
    fn permutation_arguments(&self) -> &[PermutationArgumentDescriptor];
}
