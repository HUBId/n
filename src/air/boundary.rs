//! Boundary constraint helpers for AIR definitions.
//! Provides deterministic enforcement of initial and final state conditions.

use crate::field::FieldElement;

/// Boundary constraint defined by fixed values at specific steps.
#[derive(Debug, Clone)]
pub struct BoundaryConstraint {
    /// Step index in the trace.
    pub step: usize,
    /// Expected register values at the step.
    pub values: Vec<FieldElement>,
}

impl BoundaryConstraint {
    /// Creates a new boundary constraint descriptor.
    pub fn new(step: usize, values: Vec<FieldElement>) -> Self {
        Self { step, values }
    }
}
