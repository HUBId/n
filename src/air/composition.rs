//! Constraint composition and evaluation helpers.
//!
//! The actual constraint polynomial logic will be connected as the prover and
//! verifier pipelines mature.
