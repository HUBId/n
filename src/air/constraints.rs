//! Constraint traits describing the AIR interface.
//! Applications implement these traits to define algebraic constraints.

use crate::field::FieldElement;

/// Trait representing the Algebraic Intermediate Representation for a computation.
pub trait Air {
    /// Returns the number of registers in the execution trace.
    fn trace_width(&self) -> usize;
    /// Returns the transition constraint evaluator.
    fn transition(&self) -> &dyn ConstraintEvaluator;
    /// Returns the boundary constraint evaluator.
    fn boundary(&self) -> &dyn ConstraintEvaluator;
}

/// Trait implemented by deterministic constraint evaluators.
pub trait ConstraintEvaluator {
    /// Evaluates constraints at the provided step and returns the resulting values.
    fn evaluate(&self, step: usize, registers: &[FieldElement]) -> Vec<FieldElement>;
}
