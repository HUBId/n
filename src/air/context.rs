//! Trace- und Kontextbeschreibungen fuer AIR-Instanzen.
//!
//! Dieses Modul dokumentiert, wie Spuren aufgebaut werden und welche Metadaten
//! eine konkrete [`AirSpec`](crate::air::traits::AirSpec) bereitstellen muss.
//! Die Strukturen sind rein beschreibend und enthalten keinerlei
//! Evaluierungslogik. Alle Felder folgen der Little-Endian Serialisierung.

use super::traits::TraceGroup;

/// Beschreibt die dichte Ausfuehrungsspur einer AIR-Instanz.
///
/// * `width` zaehlt die Gesamtzahl aller Registerspalten (Core + Aux + Selektoren).
/// * `steps` definiert die Spurlaenge in Schritten; Schrittweite ist strikt `1`.
/// * `registers` listet alle Register in kanonischer Reihenfolge (zuerst Core,
///   danach Aux, zuletzt Selektoren). Jeder Eintrag enthaelt Offsets, damit
///   Speicherblöcke nachvollziehbar bleiben.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TraceInfo {
    /// Gesamtbreite des Traces in Spalten.
    pub width: usize,
    /// Anzahl der Schritte in der dichten Spur.
    pub steps: usize,
    /// Kanonische Beschreibung saemtlicher Register.
    pub registers: Vec<TraceRegister>,
}

/// Beschreibt ein einzelnes Register innerhalb der Spur.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TraceRegister {
    /// Eindeutiger Index (0-basierend) innerhalb des Register-Arrays.
    pub column: usize,
    /// Gruppenzugehoerigkeit des Registers.
    pub group: TraceGroup,
    /// Byte-Offset bezogen auf den Anfang des jeweiligen Gruppenspeichers
    /// (column-major, LE-Serialisierung).
    pub group_offset: usize,
}

/// Offset der Trace-Domäne relativ zum Einheitswurzelgenerator.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct TraceDomainOffset {
    /// Rohbytes der Offset-Repräsentation im LE-Format.
    pub le_bytes: [u8; 32],
}

/// Digest einer deterministischen FRI-Planinstanz.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct FriPlanDigest {
    /// 32 Byte BLAKE2s-kompatibler Digest im LE-Format.
    pub le_bytes: [u8; 32],
}

/// Kontextbeschreibung, die jede [`AirSpec`](crate::air::traits::AirSpec)
/// zur Verfügung stellen muss.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct AirContext {
    /// Spur- und Registerinformationen in kanonischer Ordnung.
    pub trace: TraceInfo,
    /// Höchstgrad der Kompositionspolynome.
    pub composition_degree: usize,
    /// Maximalgrad einzelner Randbedingungen.
    pub boundary_degree: usize,
    /// Maximalgrad einzelner Uebergangsbedingungen.
    pub transition_degree: usize,
    /// Anzahl der verlangten OOD-Punkte (>= 2 laut Spezifikation).
    pub ood_points: usize,
    /// Faktor für die Low-Degree-Extension.
    pub lde_factor: usize,
    /// Offset der Trace-Domäne für Core- und Aux-Gruppen.
    pub domain_offset: TraceDomainOffset,
    /// Digest des FRI-Plans für die Beweisinstanz.
    pub fri_plan: FriPlanDigest,
    /// Deterministische Gruppenzuordnung für Commitments.
    pub group_digests: Vec<ParameterDigest>,
}

use super::ids::ParameterDigest;
