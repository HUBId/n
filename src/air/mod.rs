//! Algebraic Intermediate Representation (AIR) module.
//! Provides traits for defining transition and boundary constraints used in STARK proofs.

pub mod boundary;
pub mod constraints;
pub mod transition;

pub use constraints::{Air, ConstraintEvaluator};
