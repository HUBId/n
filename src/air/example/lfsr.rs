//! Linear-feedback shift register example AIR.
//!
//! This module provides a compact worked example that exercises the public
//! input codec, deterministic trace construction and constraint recording APIs
//! exposed by the AIR layer.  The example models a single-register LFSR defined
//! over the prime field used by the library.  The recurrence is intentionally
//! simple: every successive step is derived as `next = ALPHA * current + BETA`.
//! This keeps the symbolic transition polynomial linear while still covering the
//! complete pipeline (codec, trace generation, boundary checks and constraint
//! recording).

use crate::air::trace::Trace;
use crate::air::traits::{
    Air as AirTrait, BoundaryBuilder as BoundaryBuilderTrait, BoundaryConstraint, Constraint,
    Evaluator as EvaluatorTrait, PolyExpr, PublicInputsCodec as PublicInputsCodecTrait,
    TraceBuilder as TraceBuilderTrait,
};
use crate::air::types::{
    AirError, BoundaryAt, ColIx, DegreeBounds, LdeOrder, PublicFieldMeta, PublicFieldType,
    PublicSpec, SerKind, TraceColMeta, TraceData, TraceRole, TraceSchema,
};
use crate::field::prime_field::{CanonicalSerialize, FieldDeserializeError, FieldElementOps};
use crate::field::FieldElement as Felt;
use crate::utils::serialization::FieldElementBytes;

/// Column index used throughout the example.
const STATE_COL: ColIx = ColIx::new(0);
/// Multiplicative factor applied during the transition.
const ALPHA: Felt = Felt(5);
/// Additive tweak applied during the transition.
const BETA: Felt = Felt(7);
/// Blowup factor enforced by the schema.
const LDE_BLOWUP: usize = 8;

/// Public inputs for the LFSR example.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PublicInputs {
    /// Seed used as the first trace value.
    pub seed: Felt,
    /// Number of rows in the execution trace.
    pub length: usize,
}

impl PublicInputs {
    /// Creates a new public input descriptor, validating the supplied length.
    pub fn new(seed: Felt, length: usize) -> Result<Self, AirError> {
        if length < 2 {
            return Err(AirError::LayoutViolation(
                "lfsr trace length must contain at least two rows",
            ));
        }
        if length % LDE_BLOWUP != 0 {
            return Err(AirError::LayoutViolation(
                "lfsr trace length must be a multiple of the lde blowup factor",
            ));
        }
        Ok(Self { seed, length })
    }

    /// Returns the canonical 32-byte digest derived from the public inputs.
    pub fn digest(&self) -> [u8; 32] {
        let mut digest = [0u8; 32];
        let seed_bytes = self.seed.to_bytes();
        digest[..seed_bytes.len()].copy_from_slice(&seed_bytes);
        let length_fe = Felt::from(self.length as u64);
        let length_bytes = length_fe.to_bytes();
        let offset = seed_bytes.len();
        digest[offset..offset + length_bytes.len()].copy_from_slice(&length_bytes);
        digest
    }
}

/// Codec implementing canonical serialization for [`PublicInputs`].
#[derive(Debug, Default, Clone, Copy)]
pub struct PublicInputsCodec;

impl PublicInputsCodec {
    fn decode_field(bytes: &FieldElementBytes) -> Result<Felt, AirError> {
        let mut buf = [0u8; 8];
        buf.copy_from_slice(&bytes.bytes[..8]);
        if bytes.bytes[8..].iter().any(|b| *b != 0) {
            return Err(AirError::Serialization {
                kind: SerKind::PublicInput,
                detail: "non-canonical field encoding",
            });
        }
        match Felt::from_bytes(&buf) {
            Ok(value) => Ok(value),
            Err(FieldDeserializeError::FieldDeserializeNonCanonical) => {
                Err(AirError::Serialization {
                    kind: SerKind::PublicInput,
                    detail: "field element not canonical",
                })
            }
        }
    }
}

impl PublicInputsCodecTrait for PublicInputsCodec {
    type Value = PublicInputs;

    fn encode(&self, value: &Self::Value) -> Result<Vec<FieldElementBytes>, AirError> {
        let mut seed_bytes = [0u8; 32];
        let seed_le = value.seed.to_bytes();
        seed_bytes[..seed_le.len()].copy_from_slice(&seed_le);

        let mut length_bytes = [0u8; 32];
        let length_fe = Felt::from(value.length as u64);
        let length_le = length_fe.to_bytes();
        length_bytes[..length_le.len()].copy_from_slice(&length_le);

        Ok(vec![
            FieldElementBytes { bytes: seed_bytes },
            FieldElementBytes {
                bytes: length_bytes,
            },
        ])
    }

    fn decode(&self, bytes: &[FieldElementBytes]) -> Result<Self::Value, AirError> {
        if bytes.len() != 2 {
            return Err(AirError::Serialization {
                kind: SerKind::PublicInput,
                detail: "lfsr codec expects two field elements",
            });
        }
        let seed = Self::decode_field(&bytes[0])?;
        let length = Self::decode_field(&bytes[1])?;
        let length = u64::from(length) as usize;
        PublicInputs::new(seed, length)
    }
}

/// Deterministic execution trace builder for the LFSR example.
#[derive(Debug, Clone)]
pub struct TraceBuilder {
    schema: TraceSchema,
    expected_column: Vec<Felt>,
    recorded_column: Option<Vec<Felt>>,
}

impl TraceBuilder {
    fn new(schema: TraceSchema, expected_column: Vec<Felt>) -> Self {
        Self {
            schema,
            expected_column,
            recorded_column: None,
        }
    }
}

impl TraceBuilderTrait for TraceBuilder {
    fn add_column(&mut self, role: TraceRole, values: Vec<Felt>) -> Result<ColIx, AirError> {
        if role != TraceRole::Main {
            return Err(AirError::LayoutViolation(
                "lfsr example only exposes a main trace column",
            ));
        }
        if values.len() != self.expected_column.len() {
            return Err(AirError::SchemaMismatch {
                what: "trace row count",
                expected: self.expected_column.len(),
                actual: values.len(),
            });
        }
        for (step, (expected, observed)) in self.expected_column.iter().zip(&values).enumerate() {
            if expected != observed {
                return Err(AirError::NonDeterministicWitness {
                    column: STATE_COL,
                    step,
                });
            }
        }
        if let Some(existing) = &self.recorded_column {
            for (step, (lhs, rhs)) in existing.iter().zip(&values).enumerate() {
                if lhs != rhs {
                    return Err(AirError::NonDeterministicWitness {
                        column: STATE_COL,
                        step,
                    });
                }
            }
        } else {
            self.recorded_column = Some(values.clone());
        }
        Ok(STATE_COL)
    }

    fn build(self, degree_bounds: DegreeBounds) -> Result<TraceData, AirError> {
        let column = self.recorded_column.ok_or(AirError::SchemaMismatch {
            what: "trace column count",
            expected: 1,
            actual: 0,
        })?;
        degree_bounds.ensure(TraceRole::Main, column.len().saturating_sub(1))?;
        let trace_data = TraceData::new(vec![column.clone()])?;
        self.schema.validate_trace(&trace_data)?;
        let trace = Trace::from_columns(self.schema.clone(), vec![column])?;
        check_transitions(&trace, &transition_expr(STATE_COL), STATE_COL)?;
        Ok(trace_data)
    }
}

/// Boundary builder that enforces determinism via the expected trace.
#[derive(Debug, Clone)]
pub struct BoundaryBuilder {
    expected_trace: Trace,
    assignments: Vec<BoundaryConstraint>,
    seen: Vec<((ColIx, BoundaryAt), Felt)>,
}

impl BoundaryBuilder {
    fn new(expected_trace: Trace) -> Self {
        Self {
            expected_trace,
            assignments: Vec::new(),
            seen: Vec::new(),
        }
    }
}

impl BoundaryBuilderTrait for BoundaryBuilder {
    fn set(&mut self, column: ColIx, at: BoundaryAt, value: Felt) -> Result<(), AirError> {
        let expected_value = match at {
            BoundaryAt::First => self.expected_trace.row(0)?.get(column)?,
            BoundaryAt::Last => self
                .expected_trace
                .row(self.expected_trace.length().saturating_sub(1))?
                .get(column)?,
            BoundaryAt::Row(ix) => self.expected_trace.row(ix)?.get(column)?,
        };
        if value != expected_value {
            let step = match at {
                BoundaryAt::First => 0,
                BoundaryAt::Last => self.expected_trace.length().saturating_sub(1),
                BoundaryAt::Row(ix) => ix,
            };
            return Err(AirError::NonDeterministicWitness { column, step });
        }
        if let Some((_, existing)) = self
            .seen
            .iter()
            .find(|((col, boundary), _)| *col == column && *boundary == at)
        {
            if existing != &value {
                let step = match at {
                    BoundaryAt::First => 0,
                    BoundaryAt::Last => self.expected_trace.length().saturating_sub(1),
                    BoundaryAt::Row(ix) => ix,
                };
                return Err(AirError::NonDeterministicWitness { column, step });
            }
        }
        self.seen.push(((column, at), value));
        self.assignments
            .push(BoundaryConstraint { column, at, value });
        Ok(())
    }

    fn build(self) -> Result<Vec<BoundaryConstraint>, AirError> {
        Ok(self.assignments)
    }
}

/// Constraint evaluator storing transition polynomials.
#[derive(Debug, Default, Clone)]
pub struct Evaluator {
    constraints: Vec<Constraint>,
}

impl EvaluatorTrait for Evaluator {
    fn enforce_zero(&mut self, expr: PolyExpr) -> Result<(), AirError> {
        self.constraints.push(Constraint::new(expr));
        Ok(())
    }

    fn constraints(&self) -> Result<Vec<Constraint>, AirError> {
        Ok(self.constraints.clone())
    }
}

/// AIR implementation for the worked LFSR example.
#[derive(Debug, Clone)]
pub struct Air {
    public_inputs: PublicInputs,
}

impl Air {
    /// Creates a new AIR instance for the provided public inputs.
    pub fn new(public_inputs: PublicInputs) -> Self {
        Self { public_inputs }
    }

    /// Stable identifier used in transcripts and documentation.
    pub const fn id() -> &'static str {
        "LFSR_Example_v1"
    }

    fn trace_columns(&self) -> Vec<TraceColMeta> {
        vec![TraceColMeta::new("state", TraceRole::Main)
            .with_boundary(BoundaryAt::First)
            .with_boundary(BoundaryAt::Last)]
    }

    fn expected_column(&self) -> Vec<Felt> {
        generate_column(&self.public_inputs)
    }

    fn expected_trace(&self, schema: &TraceSchema) -> Result<Trace, AirError> {
        Trace::from_columns(schema.clone(), vec![self.expected_column()])
    }
}

impl AirTrait for Air {
    type TraceBuilder = TraceBuilder;
    type BoundaryBuilder = BoundaryBuilder;
    type Evaluator = Evaluator;
    type PublicInputsCodec = PublicInputsCodec;

    fn trace_schema(&self) -> Result<TraceSchema, AirError> {
        let degree = self.public_inputs.length.saturating_sub(1);
        let columns = self.trace_columns();
        let lde_order = LdeOrder::new(LDE_BLOWUP)?;
        let degree_bounds = DegreeBounds::new(degree, degree)?;
        TraceSchema::new(columns, lde_order, degree_bounds)
    }

    fn public_spec(&self) -> Result<PublicSpec, AirError> {
        let schema = self.trace_schema()?;
        let spec = PublicSpec {
            fields: vec![
                PublicFieldMeta {
                    label: "seed".to_string(),
                    field_type: PublicFieldType::TraceValue {
                        column: STATE_COL,
                        boundary: BoundaryAt::First,
                    },
                },
                PublicFieldMeta {
                    label: "final_state".to_string(),
                    field_type: PublicFieldType::TraceValue {
                        column: STATE_COL,
                        boundary: BoundaryAt::Last,
                    },
                },
            ],
        };
        spec.validate(&schema)?;
        Ok(spec)
    }

    fn new_trace_builder(&self) -> Result<Self::TraceBuilder, AirError> {
        let schema = self.trace_schema()?;
        let expected_column = self.expected_column();
        Ok(TraceBuilder::new(schema, expected_column))
    }

    fn new_boundary_builder(&self) -> Result<Self::BoundaryBuilder, AirError> {
        let schema = self.trace_schema()?;
        let expected_trace = self.expected_trace(&schema)?;
        Ok(BoundaryBuilder::new(expected_trace))
    }

    fn new_evaluator(&self) -> Result<Self::Evaluator, AirError> {
        let mut evaluator = Evaluator::default();
        evaluator.enforce_zero(transition_expr(STATE_COL))?;
        Ok(evaluator)
    }

    fn public_inputs_codec(&self) -> Result<Self::PublicInputsCodec, AirError> {
        Ok(PublicInputsCodec)
    }
}

fn generate_column(inputs: &PublicInputs) -> Vec<Felt> {
    let mut column = Vec::with_capacity(inputs.length);
    let mut state = inputs.seed;
    column.push(state);
    for _ in 1..inputs.length {
        state = next_state(state);
        column.push(state);
    }
    column
}

fn next_state(current: Felt) -> Felt {
    current.mul(&ALPHA).add(&BETA)
}

fn transition_expr(column: ColIx) -> PolyExpr {
    let linear = PolyExpr::mul_const(PolyExpr::col(column), ALPHA);
    let affine = PolyExpr::add_const(linear, BETA);
    PolyExpr::sub(PolyExpr::next(column), affine)
}

fn evaluate_poly(expr: &PolyExpr, current: Felt, next: Felt) -> Felt {
    match expr {
        PolyExpr::Const(value) => *value,
        PolyExpr::Column { .. } => current,
        PolyExpr::Next { .. } => next,
        PolyExpr::Neg(inner) => evaluate_poly(inner, current, next).neg(),
        PolyExpr::Add(lhs, rhs) => {
            evaluate_poly(lhs, current, next).add(&evaluate_poly(rhs, current, next))
        }
        PolyExpr::Sub(lhs, rhs) => {
            evaluate_poly(lhs, current, next).sub(&evaluate_poly(rhs, current, next))
        }
        PolyExpr::Mul(lhs, rhs) => {
            evaluate_poly(lhs, current, next).mul(&evaluate_poly(rhs, current, next))
        }
    }
}

fn check_transitions(trace: &Trace, expr: &PolyExpr, column: ColIx) -> Result<(), AirError> {
    if trace.length() < 2 {
        return Err(AirError::LayoutViolation(
            "lfsr trace must contain at least two rows",
        ));
    }
    for step in 0..(trace.length() - 1) {
        let pair = trace.row_pair(step)?;
        let (current, next) = pair.get(column)?;
        let value = evaluate_poly(expr, current, next);
        if !value.is_zero() {
            return Err(AirError::InvalidTransition {
                column,
                step,
                current,
                next,
            });
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_json_snapshot;
    use proptest::prelude::*;

    #[test]
    fn codec_roundtrip() {
        let inputs = PublicInputs::new(Felt(9), 8).unwrap();
        let codec = PublicInputsCodec;
        let encoded = codec.encode(&inputs).unwrap();
        let decoded = codec.decode(&encoded).unwrap();
        assert_eq!(decoded, inputs);
    }

    #[test]
    fn codec_rejects_non_canonical_encoding() {
        let codec = PublicInputsCodec;
        let mut non_canonical = FieldElementBytes { bytes: [0u8; 32] };
        non_canonical.bytes[8] = 1;
        let err = codec
            .decode(&[non_canonical.clone(), non_canonical])
            .unwrap_err();
        assert!(matches!(
            err,
            AirError::Serialization {
                kind: SerKind::PublicInput,
                detail
            } if detail.contains("non-canonical")
        ));
    }

    #[test]
    fn trace_generation_is_deterministic() {
        let inputs = PublicInputs::new(Felt(11), 8).unwrap();
        let air = Air::new(inputs.clone());
        let schema = air.trace_schema().unwrap();
        let degree_bounds = schema.degree_bounds;
        let expected_column = generate_column(&inputs);

        let mut builder = air.new_trace_builder().unwrap();
        builder
            .add_column(TraceRole::Main, expected_column.clone())
            .unwrap();
        let trace_data = builder.build(degree_bounds).unwrap();
        assert_eq!(trace_data.num_columns(), 1);

        let mut second_builder = air.new_trace_builder().unwrap();
        second_builder
            .add_column(TraceRole::Main, expected_column.clone())
            .unwrap();
        let second_trace_data = second_builder.build(schema.degree_bounds).unwrap();
        assert_eq!(
            trace_data.column(STATE_COL).unwrap(),
            second_trace_data.column(STATE_COL).unwrap()
        );

        let mut mismatched = expected_column.clone();
        mismatched[3] = mismatched[3].add(&Felt::ONE);
        let mut failing_builder = air.new_trace_builder().unwrap();
        let err = failing_builder
            .add_column(TraceRole::Main, mismatched)
            .unwrap_err();
        assert!(matches!(err, AirError::NonDeterministicWitness { .. }));
    }

    #[test]
    fn public_input_digest_snapshot() {
        let inputs = PublicInputs::new(Felt(3), 8).unwrap();
        assert_json_snapshot!("public_input_digest_seed3_len8", inputs.digest());
    }

    #[test]
    fn transition_violation_is_detected() {
        let inputs = PublicInputs::new(Felt(13), 8).unwrap();
        let air = Air::new(inputs.clone());
        let schema = air.trace_schema().unwrap();
        let mut column = generate_column(&inputs);
        column[4] = column[4].add(&Felt::ONE);
        let trace = Trace::from_columns(schema, vec![column]).unwrap();
        let err = check_transitions(&trace, &transition_expr(STATE_COL), STATE_COL).unwrap_err();
        assert!(matches!(err, AirError::InvalidTransition { .. }));
    }

    #[test]
    fn boundary_tampering_is_detected() {
        let inputs = PublicInputs::new(Felt(5), 8).unwrap();
        let air = Air::new(inputs.clone());
        let mut builder = air.new_boundary_builder().unwrap();
        builder
            .set(STATE_COL, BoundaryAt::First, inputs.seed)
            .unwrap();
        builder
            .set(
                STATE_COL,
                BoundaryAt::Last,
                generate_column(&inputs).last().copied().unwrap(),
            )
            .unwrap();
        builder.build().unwrap();

        let mut tampered = air.new_boundary_builder().unwrap();
        let err = tampered
            .set(STATE_COL, BoundaryAt::First, inputs.seed.add(&Felt::ONE))
            .unwrap_err();
        assert!(matches!(err, AirError::NonDeterministicWitness { .. }));
    }

    #[test]
    fn trace_data_boundary_checks_enforced() {
        let inputs = PublicInputs::new(Felt(17), 8).unwrap();
        let air = Air::new(inputs.clone());
        let schema = air.trace_schema().unwrap();
        let expected = generate_column(&inputs);
        let mut builder = air.new_trace_builder().unwrap();
        builder
            .add_column(TraceRole::Main, expected.clone())
            .unwrap();
        let trace_data = builder.build(schema.degree_bounds).unwrap();

        let last = expected.last().copied().unwrap();
        assert_eq!(
            trace_data
                .boundary_value(STATE_COL, BoundaryAt::First)
                .unwrap(),
            inputs.seed
        );
        assert_eq!(
            trace_data
                .boundary_value(STATE_COL, BoundaryAt::Last)
                .unwrap(),
            last
        );

        let err = trace_data
            .boundary_value(STATE_COL, BoundaryAt::Row(expected.len() + 1))
            .unwrap_err();
        assert!(matches!(err, AirError::BoundaryViolation { .. }));
    }

    proptest! {
        #![proptest_config(ProptestConfig::with_cases(32))]
        #[test]
        fn proptest_trace_is_deterministic(seed in 0u64..32, length_factor in 1usize..4) {
            let length = (length_factor + 1) * LDE_BLOWUP;
            let inputs = PublicInputs::new(Felt(seed), length).unwrap();
            let air = Air::new(inputs.clone());
            let schema = air.trace_schema().unwrap();
            let expected_column = generate_column(&inputs);

            let mut builder_a = air.new_trace_builder().unwrap();
            builder_a.add_column(TraceRole::Main, expected_column.clone()).unwrap();
            let trace_a = builder_a.build(schema.degree_bounds).unwrap();

            let mut builder_b = air.new_trace_builder().unwrap();
            builder_b.add_column(TraceRole::Main, expected_column.clone()).unwrap();
            let trace_b = builder_b.build(schema.degree_bounds).unwrap();

            prop_assert_eq!(trace_a.column(STATE_COL).unwrap(), trace_b.column(STATE_COL).unwrap());

            if length > 1 {
                let mut tampered = expected_column.clone();
                let ix = (seed as usize) % length;
                tampered[ix] = tampered[ix].add(&Felt::ONE);
                let mut tampered_builder = air.new_trace_builder().unwrap();
                let err = tampered_builder.add_column(TraceRole::Main, tampered).unwrap_err();
                let witness_violation = matches!(err, AirError::NonDeterministicWitness { .. });
                prop_assert!(witness_violation);
            }

            let mut boundary_builder = air.new_boundary_builder().unwrap();
            boundary_builder.set(STATE_COL, BoundaryAt::First, inputs.seed).unwrap();
            let last = expected_column.last().copied().unwrap();
            boundary_builder.set(STATE_COL, BoundaryAt::Last, last).unwrap();
            boundary_builder.build().unwrap();
        }
    }
}
