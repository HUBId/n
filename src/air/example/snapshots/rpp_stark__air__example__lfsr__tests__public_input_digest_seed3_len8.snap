---
source: src/air/example/lfsr.rs
assertion_line: 502
expression: inputs.digest()
---
[
  3,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  8,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0
]

